airbyte-lib/airbyte_lib/caches/base.py

```diff
@@ -0,0 +1,591 @@
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+
+"""A SQL Cache implementation."""
+
+import abc
+import enum
+from collections.abc import Generator, Iterable
+from contextlib import contextmanager
+from functools import cached_property, lru_cache
+from pathlib import Path
+from typing import TYPE_CHECKING, Any, cast, final
+
+import pandas as pd
+import pyarrow as pa
+import sqlalchemy
+import ulid
+from airbyte_lib.config import CacheConfigBase
+from airbyte_lib.file_writers import FileWriterBase, FileWriterBatchHandle
+from airbyte_lib.processors import BatchHandle, RecordProcessor
+from airbyte_lib.types import SQLTypeConverter
+from airbyte_protocol.models import ConfiguredAirbyteCatalog, ConfiguredAirbyteStream
+from overrides import overrides
+from sqlalchemy import ClauseElement, CursorResult, Executable, MetaData, Table, TextClause, create_engine, text
+from sqlalchemy.engine import Engine
+from sqlalchemy.pool import StaticPool
+
+if TYPE_CHECKING:
+    from sqlalchemy.engine import Connection
+    from sqlalchemy.engine.reflection import Inspector
+
+
+DEBUG_MODE = True
+
+
+class RecordDedupeMode(enum.Enum):
+    APPEND = "append"
+    REPLACE = "replace"
+
+
+class SQLRuntimeError(Exception):
+    """Raised when an SQL operation fails."""
+
+
+class SQLCacheConfigBase(CacheConfigBase):
+    """Same as a regular config except it exposes the 'get_sql_alchemy_url()' method."""
+
+    dedupe_mode = RecordDedupeMode.APPEND
+    schema_name: str = "airbyte_raw"
+    table_prefix: str = ""
+    table_suffix: str = ""
+
+    @abc.abstractmethod
+    def get_sql_alchemy_url(self) -> str:
+        """Returns a SQL Alchemy URL."""
+        ...
+
+    @abc.abstractmethod
+    def get_database_name(self) -> str:
+        """Return the name of the database."""
+        ...
+
+
+class GenericSQLCacheConfig(SQLCacheConfigBase):
+    """Allows configuring 'sql_alchemy_url' directly."""
+
+    sql_alchemy_url: str
+
+    @overrides
+    def get_sql_alchemy_url(self) -> str:
+        """Returns a SQL Alchemy URL."""
+        return self.sql_alchemy_url
+
+
+class SQLCacheBase(RecordProcessor):
```

Comment:
Per the spec, the cache should advertise all streams it knows about via `streams.items()`:
```
for stream_name, stream in cache.streams.items():
```

Seems like this part is missing (this is related to the comment in the examples which don't work anymore)

Reply:
I've added `streams` for caches, although for now it is not implemented.